src/main/java/com/limechain/trie/DiskTrieService.java

```diff
@@ -0,0 +1,826 @@
+import com.limechain.trie.structure.node.TrieNodeData;
+import com.limechain.utils.HashUtils;
+import lombok.Getter;
+import lombok.extern.java.Log;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+@Log
+public class DiskTrieService {
+
+    private final TrieStorage trieStorage;
+    public final TrieChanges trieChanges;
+
+    public DiskTrieService(TrieStorage trieStorage) {
+        this.trieStorage = trieStorage;
+
+        this.trieChanges = TrieChanges.empty();
+    }
+
+    public Optional<byte[]> findStorageValue(byte[] merkleRoot, Nibbles key) {
+        Optional<PendingTrieNodeChange> change = trieChanges.getFromCache(key);
+        return change.<Optional<byte[]>>map(pendingTrieNodeChange ->
+                pendingTrieNodeChange instanceof PendingInsertUpdate update
+                    ? Optional.ofNullable(update.value())
+                    : Optional.empty())
+            .orElseGet(() -> traverseTrie(merkleRoot, key) instanceof TraversalResult.Found found
+                ? Optional.ofNullable(found.getFound().getValue())
+                : Optional.empty());
+    }
+
+    public Optional<Nibbles> getNextKey(byte[] merkleRoot, Nibbles key) {
+        PendingInsertUpdate cachedRoot = trieChanges.getRoot().orElse(null);
+        TrieNodeData root = cachedRoot != null
+            ? new TrieNodeData(cachedRoot.value() != null,
+            cachedRoot.partialKey(),
+            cachedRoot.childrenMerkleValues(),
+            cachedRoot.value(),
+            null,
+            (byte) cachedRoot.stateVersion().asInt())
+            : trieStorage.getTrieNodeFromMerkleValue(merkleRoot);
+
+        if (root == null) {
+            return Optional.empty();
+        }
+
+        return findNextKey(key, Nibbles.EMPTY, root);
+    }
+
+    private Optional<Nibbles> findNextKey(Nibbles prefix, Nibbles currentKey, TrieNodeData node) {
+        if (node.getValue() != null && currentKey.compareTo(prefix) > 0) {
+            return Optional.of(currentKey);
+        }
+
+        for (Nibble nibble : Nibbles.ALL) {
+            byte[] childMerkle = node.getChildrenMerkleValues().get(nibble.asInt());
+            if (childMerkle == null) {
+                continue;
+            }
+
+            TrieNodeData cachedChild = getCachedChildAtIndex(currentKey, nibble);
+            TrieNodeData childNode = cachedChild != null
+                ? cachedChild
+                : trieStorage.getTrieNodeFromMerkleValue(childMerkle);
+
+            Nibbles nextPath = currentKey.add(nibble).addAll(childNode.getPartialKey());
+            Optional<Nibbles> result = findNextKey(prefix, nextPath, childNode);
+            if (result.isPresent()) {
+                return result;
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public void upsertNode(byte[] merkleRoot, Nibbles key, byte[] storageValue, StateVersion stateVersion) {
```

Comment:
What does 'upsert' mean?

Reply:
It's usually a database operation meaning "update/insert". I find it appropriate for cases like this one where the method can do both operations depending on some underlying logic.